sound/soc/sof/ipc4-loader.c

```diff
@@ -256,26 +262,80 @@ static int sof_ipc4_load_library_by_uuid(struct snd_sof_dev *sdev,
 	if (ret)
 		goto release;
 
-	ret = xa_insert(&ipc4_data->fw_lib_xa, lib_id, fw_lib, GFP_KERNEL);
+	ret = xa_insert(&ipc4_data->fw_lib_xa, *lib_id, fw_lib, GFP_KERNEL);
 	if (unlikely(ret))
 		goto release;
 
-	kfree(fw_filename);
-
+	(*lib_id)++;
 	return 0;
 
 release:
 	release_firmware(fw_lib->sof_fw.fw);
 	/* Allocated within sof_ipc4_fw_parse_ext_man() */
 	devm_kfree(sdev->dev, fw_lib->modules);
-free_filename:
-	kfree(fw_filename);
 free_fw_lib:
 	devm_kfree(sdev->dev, fw_lib);
 
 	return ret;
 }
 
+int sof_ipc4_load_library_bundles(struct snd_sof_dev *sdev)
+{
+	static const char * const lib_bundle[] = { "openmodules", "debug" };
+	const char *fw_filename = sdev->pdata->fw_filename;
+	const char *lib_filename, *p;
+	unsigned long lib_id = 1;
+	char *lib_name_base;
+	int ret, i;
+
+	p = strstr(fw_filename, ".ri");
+	if (!p || strlen(p) != 3) {
+		dev_info(sdev->dev,
+			 "%s: Cannot parse firmware name '%s', missing .ri extension\n",
+			 __func__, fw_filename);
+		return 0;
+	}
+
+	lib_name_base = kzalloc(strlen(fw_filename) - 2, GFP_KERNEL);
+	strscpy(lib_name_base, fw_filename, sizeof(lib_name_base));
+
+	for (i = 0; i < ARRAY_SIZE(lib_bundle); i++) {
+		lib_filename = kasprintf(GFP_KERNEL, "%s/%s-%s.ri",
+					 sdev->pdata->fw_lib_prefix,
+					 lib_name_base, lib_bundle[i]);
+		if (!lib_filename)
+			return -ENOMEM;
+
+		ret = sof_ipc4_load_library(sdev, &lib_id, lib_filename, true);
+
+		kfree(lib_filename);
+		if (ret)
+			break;
```

Comment:
why `break`? If any of the libraries fail loading you don't attempt the rest? In this case it kind of makes sense - if there's no "main" module bundle, there's probably no "debug" either, but isn't the general concept - here's a list of optional libraries, we try to load whatever is available?

Reply:
We will fail to probe the card as the modules form the failed library are missing and the topology will not probe.

Reply:
OK, if we continue then the car will not come up, but the debugfs for SOF is going the be there to provide further clues.